climakitae/uncertain_utils.py

```diff
@@ -420,3 +617,135 @@ def compute_vmin_vmax(da_min, da_max):
     else:
         sopt = None
     return vmin, vmax, sopt
+
+
+def get_ks_pval_df(sample1, sample2, sig_lvl=0.05):
+    """Performs a Kolmogorov-Smirnov test at all lat, lon points
+
+    Parameters
+    ----------
+    sample1: xr.Dataset
+        first sample for comparison
+    sample2: xr.Dataset
+        sample against which to compare sample1
+    sig_lvl: float
+        alpha level for statistical significance
+
+    Returns
+    -------
+    pd.DataFrame
+        columns are lat, lon, and p_value;
+        only retains spatial points where
+        p_value < sig_lvl
+    """
+
+    sample1 = sample1.stack(allpoints=["y", "x"]).squeeze().groupby("allpoints")
+    sample2 = sample2.stack(allpoints=["y", "x"]).squeeze().groupby("allpoints")
+
+    def ks_stat_2sample(sample1, sample2):
+        try:
+            ks = stats.kstest(sample1, sample2)
+            d_statistic = ks[0]
+            p_value = ks[1]
+        except (ValueError, ZeroDivisionError):
+            d_statistic = np.nan
+            p_value = np.nan
+
+        return d_statistic, p_value
+
+    d_statistic, p_value = xr.apply_ufunc(
+        ks_stat_2sample,
+        sample1,
+        sample2,
+        input_core_dims=[["index"], ["index"]],
+        exclude_dims=set(("index",)),
+        output_core_dims=[[], []],
+    )
+
+    p_df = p_value.dropna(dim="allpoints")
+    p_df = p_value.rename("p_value")
+    p_df = p_df.unstack("allpoints")
+    p_df = p_df.to_dataframe().reset_index()
+    p_df = p_df[["lat", "lon", "p_value"]]
+    p_df = p_df.loc[:, ["lon", "lat", "p_value"]]
+    p_df = p_df[p_df["p_value"] < sig_lvl]
+
+    return p_df
+
+
+def get_warm_level(warm_level, ds, multi_ens=False, ipcc=True):
+    """Subsets projected data centered to the year
+    that the selected warming level is reached
+    for a particular simulation/member_id
+
+    Parameters
+    ----------
+    warm_level: float or int
+        options: 1.5, 2.0, 3.0, 4.0
+    ds: xr.Dataset
+        Can only have one 'simulation' coordinate
+    multi_ens: bool, default=False
+        Set to True if passing a simulation with multiple member_id
+    ipcc: bool, default=True
+        Set to False if performing warming level analysis with
+        respect to IPCC standard baseline (1850-1900)
+
+    Returns
+    -------
+    xr.Dataset
+        Subset of projected data -14/+15 years from warming level threshold
+    """
+    wtype = type(warm_level)
```

Comment:
This can replace line 698...704

```
try:
    warm_level = float(warm_level)
except ValueError:
    raise Exception("Please specify warming level as an integer or float.")
```

the `float` function will throw an error for us if it can't convert or is out of range. We are just intercepting the `ValueError` exception and adding our own explanation.

Reply:
ah thanks! fixing now